src/hoisted.jl

```diff
@@ -0,0 +1,84 @@
+# Specialized functions for increased performance when JSON is in-memory
+
+addbyte!(x::Int, c) = x + 1
+addbyte!(x::Vector{UInt8}, c) = push!(x, c)
+
+function parse_string(ps::MemoryParserState)
+    # "Dry Run": find length of string so we can allocate the right amount of
+    # memory from the start.
+    s = ps.s
+    len = parse_string(ps, 0)
+
+    # Now read the string itself
+    ps.s = s
+    ret = Compat.UTF8String(UInt8[])
+    b = sizehint!(ret.data, len)
+    parse_string(ps, b)
+    ret
+end
+
+@inbounds function parse_string(ps::MemoryParserState, b)
+    str = ps.utf8data   # manual hoisting
+    s = ps.s
+    e = endof(str)
+
+    s += 1  # skip past opening string character "
+    while s <= e
+        c = str[s]
+        if c == BACKSLASH
+            s += 1
+            s > e && break
+            c = str[s]
+            if c == LATIN_U  # Unicode escape
+                ps.s = s + 1
+                for c in string(read_unicode_escape!(ps)).data
+                    b = addbyte!(b, c)
+                end
+                s = ps.s
+                continue
+            elseif haskey(ESCAPES, c)
+                c = ESCAPES[c]
```

Comment:
could avoid the double hash-table lookup here by `c = get(ESCAPES, c, 0xff)` and then throwing an error on `c == 0xff`